### ISSUE TITLE
`labeled_counter` metrics use sample counts instead of counter values

### ISSUE THREAD

**Issue by omansfeld** [mozilla/glam#3524]
## The Bug

While looking at the `networking.http_3_congestion_event_reason` labeled_counter (labels: `loss`, `ecn_ce`) I noticed a significant discrepancy between GLAMs view (`~16% ecn_ce` earlier this year) vs what I remembered (`~3% ecn_ce` at the time).

See https://glam.telemetry.mozilla.org/fog/probe/networking_http_3_congestion_event_reason/explore?activeBuckets=%5B%22loss%22%2C%22ecn-ce%22%5D&timeHorizon=ALL vs the screenshot below that I used in a talk earlier this year:

<img width="1197" height="677" alt="Image" src="https://github.com/user-attachments/assets/553fec40-7974-4dec-8e12-b59e9177bc77" />

## Investigation

I looked into this and it seems like GLAM now uses sample counts instead of the actual counter values when calculating percentages.

I.e. a client sending a ping with `loss = 9` and `ecn_ce = 1` has one sample for each, thus it records it as `50% loss` and `50% ecn_ce` instead of the real `9/1 -> 90%/10%` proportion.

This is also visible in those two (ai assisted) queries that I wrote to understand what is happening (and what the data actually looks like):

1. Using actual proportions (in a hacky way because I tried to use the glam aggregate table): https://sql.telemetry.mozilla.org/queries/121187#295620

<img width="1161" height="530" alt="Image" src="https://github.com/user-attachments/assets/a1b591da-5efc-4e68-9a1b-b86f1606fa4a" />

versus

2. trying to mirror what GLAM seems to do (using client counts): https://sql.telemetry.mozilla.org/queries/121188#295622

<img width="1161" height="530" alt="Image" src="https://github.com/user-attachments/assets/a2cdd665-3a51-4f7b-9d06-66f192ba8870" />

## Impact

I assume this affects all labeled counters on GLAM. I have checked a few others where I remember rough values and they all don't match what I expected them to show.

I guess (hope) this is a bug and not intended. At least for me GLAM is not useful for labeled counters with this. And I am in general a big fan and try to use GLAM over STMO (and design metrics so they look useful on GLAM) whenever possible.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-11238)

**Comment by edugfilho** [comment]
This is intended per https://mozilla-hub.atlassian.net/browse/DENG-7764.
The good news is, it's not a hard change to make. @omansfeld would it help if I added a dropdown where you could pick the aggregation type? The right aggregation varies per metric. In the case of `networking_http_3_congestion_event_reason` I believe `sum` would be the right one to use.

**Comment by omansfeld** [comment]
Hey, thanks for the quick answer. 

It sounds unintuitive for me to ignore the counter values for `labeled_counter` and the approach discussed in the JIRA ticket where client A submits `{label1: 1}, {label2: 1}` and client B submits `{label1: 10}` and we get `{label1: 1.5}, label2: 0.5}` seems more like Client ID normalization to me, which as far as I can tell is already it's own option in the top row of dropdowns.

Anyways, I might be lacking context here and only have the narrow view of how we use labeled counters in my scope. But imo the dangerous thing is that the current aggregation looks correct without very in depth knowledge about how we record the metric in question or remembering it's historical values. Which then can lead to very wrong assumptions (or a lot of wasted time trying to get custom queries and the GLAM display to match in my case).

**Having a dropdown would definitely help** and then I think to some extend the person viewing GLAM has the responsibility to check that they are using the correct options for the data they are viewing (or ask an expert if unsure), so in the grand scheme of things it would probably be fine.

**Comment by edugfilho** [comment]
@omansfeld thank you for that input. I dug for more context and here are a few points:

* I realized adding the dropdown won't help: `sum` is already what's being used... exposing avg/max etc would only produce other per-client weightings, none more correct as a category comparison.

* Why we don't tend to use raw SUM of values for the proportions: A value sum is frequently dominated by a few hyper-active clients, so the bars stop reflecting the typical client. E.g. 1M clients each report category A once, while 10 stuck clients report B 1M× each → SUM says "B = 91%" when really ~all clients are in A.

* What I fixed (normalized view): The categorical % was being normalized by sample/ping count, not clients. So chatty clients were over-weighted. It now uses distinct clients per category (total_users(key) / Σ total_users), which is robust to activity volume and keeps the bars summing to 100%. The y-axis is also pinned to 0–100% with whole-percent ticks.

* Your expected behavior was also introduced, now under "non-normalized.": The sum-of-values-by-category view you were after is now what the non-normalized toggle shows. Heads up that it's an approximation (reconstructed from bucketed data, so coarsest in the tail) and outlier-sensitive... which is why it's not the default view.

Ready for review: https://github.com/mozilla/glam/pull/3525

**Comment by omansfeld** [comment]
Got it, thanks! The raw sums being bad for the reasons you outlined definitely makes a lot of sense. We probably need some type of per client normalization.

One question for me to better understand the default view I'll be seeing and to make sure I properly articulated my point: Is the normalization done over the *label sums per client*, or do we just count *label occurrence per client*?

I.e if we have
```
clientA: {label1: 9, label2: 1},
clientB: {label1: 1, label2: 1}
```
do these normalize to **option A**:
```
{label1: 2 occurrences}, [label2: 2 occurrences} --> 50% each
```
or **option B**:
```
clientA: {90%, 10%}, clientB: {50%, 50%} 
--> label1 = (0.9+0.5)/2 = 0.7
--> label2 = (0.1+0.5)/2 = 0.3
```

Both ways are accounting and normalizing for your example of a few clients submitting a very high `{label1: 10mil)`, but only **option B** would correctly display if most clients were to report `{label1: 9, label2: 1}` or a similar label distribution.

That was the case for my metric `networking_http_3_congestion_event_reason`. Most clients reported many `loss` and only a few clients reported `ecn_ce`. After we fixed a loss detection bug now most clients should either report `loss` or `ecn` or in rare cases multiple of each or a mixture. But that change wasn't visible in GLAM because (afaict) a client with `{loss: 10, ecn_ce: 1}` got normalized to `50% each`, which I think is misleading.

**Comment by edugfilho** [comment]
Yea, unfortunately it's doing **A** now. 
And I see it can get misleading. **B** is the true client normalization (the type we do with histograms), but it isn't going to be straightforward to implement, because it involves ETL changes and potentially a big backfill.

In case you're curious: To implement **B** I'll need to make changes in the ETL that account specifically for labeled counters that are categorical, which currently follow the same path as non-categorical labeled_counters, and that path buckets each label independently and collapses clients into counts early on, thus removing the possibility of computing a per-client cross-label ratio.
But that's the easy part. Where it gets more complicated is the fact I'll probably have to backfill a ton of data, unless we're ok with having a chart that shows a different type of aggregation from a certain point (I'm interested to hear your thoughts on that btw)... In any case, it's a thing I'll have to discuss with people. But given that that's the original intention of [that Jira ticket](https://mozilla-hub.atlassian.net/browse/DENG-7764) and I feel bad for having cut corners there (maybe we wouldn't need a backfill then?), I'm willing to do this now-ish.

**Comment by omansfeld** [comment]
For me having a different aggregation from a certain point wouldn't be that much a problem, or in other words, I'd rather have proper normalization in the future and a break in metrics history than not having proper normalization at all.

But I see how it would be confusing for people who don't know about the change when they look at GLAM and then think something for their actual metric changed, when it is just the aggregation. Maybe there could be a marker in GLAM in the histogram for all labeled counters that says something along the lines of "changed aggregation method" with a red line or something like that.

Similar to the below from [this](https://performance.mozilla.org/networking.html) dashboard:

<img width="374" height="589" alt="Image" src="https://github.com/user-attachments/assets/7f935249-8f68-43f6-9fe8-0ba3a68596fd" />

I don't know if you have the plumbing to easily do something like that.

Anyways, it would be cool to see labeled counters on GLAM do the normalization like in example **B** in the future! Thanks!

**Comment by edugfilho** [comment]
Thanks for that input. I'm implementing proper normalization in the ETL, and yes, I also think [adding markers on GLAM ](https://github.com/mozilla/glam/issues/3528) would be great, for many reasons.

**Comment by edugfilho** [comment]
GH automation closed this. Reopening, because but I'll manually close once all PRs related to this issue are merged and the aggregation showing on GLAM

**Comment by edugfilho** [comment]
GLAM ETL now applies proper client normalization, which can be seen on GLAM. Users will notice a drop in client count during this transition time for static `labeled_counters`, because the ETL cumulatively updates the last 3 versions of data on each run, and it's dropping the accumulated data that's been previously aggregated the wrong way, since there's no way to merge both types of aggregations.
